packages/server/src/modules/permissions/service.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  Activity,
+  Chapter,
+  hasPermissions,
+  Lesson,
+  Step,
+  TYPE_PERMISSION,
+  User,
+} from '@chess-tent/models/src';
+import { DepopulatedPermission, PermissionModel } from './model';
+import { v4 as uuid } from 'uuid';
+import { FilterQuery } from 'mongoose';
+import { AppDocument } from '@types';
+import { Permission } from '@chess-tent/models';
+
+export const addPermissionService = async (
+  object: Step | Chapter | Lesson | Activity,
+  role: string,
+  holder: User,
+) => {
+  await new PermissionModel({
+    id: uuid(),
+    type: TYPE_PERMISSION,
+    holder: holder.id,
+    holderType: holder.type,
+    role: role,
+    object: object.id,
+    objectType: object.type,
+  } as DepopulatedPermission).save();
+};
+
+export const hasPermissionToDoService = async (
+  subject: User,
+  object: Step | Chapter | Lesson | Activity, // | UserGroup
+  predicate: string,
+): Promise<boolean> => {
+  const query: FilterQuery<AppDocument<DepopulatedPermission>> = {
+    object: object.id,
+    objectType: object.type,
+    holder: subject.id,
+    holderType: subject.type,
+  };
+
+  // Once UserGroups are implemented, we'll have to
+  // filter permissions with all the userGroups user
+  // is assigned to
+  const permissions = await PermissionModel.find(query).lean().exec();
+  const roles = permissions.map(item => item.role);
+  return new Promise<boolean>(() =>
+    hasPermissions(roles, object.type, predicate),
+  );
+};
+
+export const getUsersWithRoleService = async (
+  object: Step | Chapter | Lesson | Activity, // | UserGroup
```

Comment:
I'd rather say `object: Subject` to be more generic. Then, if there is a permission with such combination it will return it.

This way we'll have to extend the service each time we add a new subject type.

Reply:
Or even change args to `id, type, role`. Idk

Reply:
There's definitely more work to do to make this strongly and openly typed.

Reply:
I'm planning to remove uses of magic text when prompting for role or action.